addons/mail/wizard/mail_resend_message.py

```diff
@@ -0,0 +1,84 @@
+
+from odoo import _, api, fields, models
+from odoo.tools.safe_eval import safe_eval
+from odoo.exceptions import UserError
+
+class MailResendMessage(models.TransientModel):
+    _name = 'mail.resend.message'
+    _description = 'Email resend wizard'
+
+    mail_message_id = fields.Many2one('mail.message', 'Message', readonly=True)
+    partner_ids = fields.One2many('mail.resend.partner', 'resend_wizard', string='Recipients')
+    notification_ids = fields.Many2many('mail.notification', string='Notifications', readonly=True)
+    has_cancell = fields.Boolean(compute='_compute_has_cancell')
+
+    @api.depends("partner_ids")
+    def _compute_has_cancell(self):
+        self.has_cancell = self.partner_ids.filtered(lambda p: not p.resend)
+
+    @api.model
+    def default_get(self, fields):
+        rec = super(MailResendMessage, self).default_get(fields)
+        message_id = self._context.get('mail_message_to_resend')
+        if message_id:
+            mail_message_id = self.env['mail.message'].browse(message_id)
+            notification_ids = mail_message_id.notification_ids.filtered(lambda notif: notif.email_status in ('exception', 'bounce'))
+            partner_readonly = not self.env['res.partner'].check_access_rights('write', raise_exception=False)
+            partner_ids = [(0, 0,
+                {
+                    "partner_id": notif.res_partner_id.id,
+                    "name": notif.res_partner_id.name,
+                    "email": notif.res_partner_id.email,
+                    "resend": True,
+                    "message": notif.format_failure_reason(),
+                    "partner_readonly": partner_readonly
+                }
+                ) for notif in notification_ids]
+            rec['notification_ids'] = [(6, 0, notification_ids.ids)]
+            rec['mail_message_id'] = mail_message_id.id
+            rec['partner_ids'] = partner_ids
+        else:
+            raise UserError('No message_id found in context')
+        return rec
+
+    @api.multi
+    def resend_mail_action(self):
+        """ Process the wizard content and proceed with sending the related
+            email(s), rendering any template patterns on the fly if needed. """
+        for wizard in self:
+            "If a partner disappeared from partner list, we cancel the notification"
+            to_cancell = wizard.partner_ids.filtered(lambda p: not p.resend).mapped("partner_id")
+            to_send = wizard.partner_ids.filtered(lambda p: p.resend).mapped("partner_id")
+            notif_to_cancell = wizard.notification_ids.filtered(lambda notif: notif.res_partner_id in to_cancell and notif.email_status in ('exception', 'bounce'))
+            notif_to_cancell.sudo().write({'email_status': 'canceled'})
+            to_send.sudo()._notify(self.mail_message_id, self.mail_message_id.layout, force_send=True, send_after_commit=False, values=safe_eval(self.mail_message_id.layout_values))

```

Comment:
use `json.loads()` unless there is a good reason to use repr/safe_eval, and in that case document it